### PR TITLE
Set/unset mbr done, enable/disable shadow mbr and write shadow mbr in libsed

### PR DESCRIPTION
**Pull request by RevanthRajashekar**

### CODE REVIEW

File: src/lib/sed.c [hunk 1]
```diff
@@ -87,6 +90,7 @@ static struct opal_interface opal_if = {
 	.set_pwd_fn = sedopal_setpw,
 	.shadow_mbr_fn = sedopal_shadowmbr,
 	.mbr_done_fn = sedopal_mbrdone,
+	.write_shadow_mbr = NULL,
```
Comment (RevanthRajashekar) [inline src/lib/sed.c:93]: Should be write_shadow_mbr_fn, this should be changed before merging

File: src/lib/sed.c [hunk 2]
```diff
@@ -111,6 +115,7 @@ static struct opal_interface opal_if = {
 	.lock_unlock_fn	= opal_lock_unlock_pt,
 	.set_pwd_fn = opal_set_pwd_pt,
 	.shadow_mbr_fn	= opal_shadow_mbr_pt,
+	.write_shadow_mbr = opal_write_shadow_mbr_pt,
```
Comment (RevanthRajashekar) [inline src/lib/sed.c:118]: Should be write_shadow_mbr_fn, this should be changed before merging

**Comment by AndrzejJakowski** [comment]
Thx Revant. Couple of comments:
1. Please change commit title so it accureately represents module that has been change (libsed insted of sedcli)
2. Split second commit into 2 patches as it represent 2 separate logical changes (generailzation of read/write table functionality and using generalized interface to write to MBR table)

File: src/lib/nvme_pt_ioctl.c [hunk 3]
```diff
@@ -2006,53 +2015,73 @@ int opal_set_pwd_pt(struct sed_device *dev, enum SED_AUTHORITY auth, const struc
...
-	}
+	if (mbr_done)
+		done = OPAL_TRUE;
+	else
+		done = OPAL_FALSE;
 
-	en_dis = (enable_disable == OPAL_MBR_ENABLE) ? OPAL_TRUE :
-		OPAL_FALSE;
+	ret = opal_start_admin1_lsp_session(dev->fd, opal_dev, key);
+        if (ret)
```
Comment (AndrzejJakowski) [inline src/lib/nvme_pt_ioctl.c]: please fix coding style issues in all the places: ERROR: code indent should use tabs where possible

File: src/lib/nvme_pt_ioctl.c [hunk 4]
```diff
@@ -2271,3 +2292,27 @@ int opal_list_lr_pt(struct sed_device *dev, const struct sed_key *key,
...
+	opal_dev = dev->priv;
+
+	ret = opal_start_admin1_lsp_session(dev->fd, opal_dev, key);
+	if (ret)
+		goto end_sessn;
+
+	ret = opal_write_mbr(dev->fd, opal_dev, from, offset, size);
+
+end_sessn:
+        opal_end_session(dev->fd, opal_dev);
```
Comment (AndrzejJakowski) [inline src/lib/nvme_pt_ioctl.c]: please fix coding style issues: ERROR: code indent should use tabs where possible

File: src/lib/nvme_pt_ioctl.c [hunk 5]
```diff
@@ -2006,53 +2015,73 @@ int opal_set_pwd_pt(struct sed_device *dev, enum SED_AUTHORITY auth, const struc
...
+	if (key == NULL) {
+		SEDCLI_DEBUG_MSG("Need ADMIN1 password for mbr shadow "\
+				"enable/disable\n");
+		return -EINVAL;
 	}
 
-	ret = opal_start_admin1_lsp_session(dev->fd, opal_dev, &disk_key);
+	opal_dev = dev->priv;
+
+	if (mbr)
```
Comment (AndrzejJakowski) [inline src/lib/nvme_pt_ioctl.c]: you could make it more pretty:)
done = mbr_done ? OPAL_TRUE : OPAL_FALSE;

File: src/lib/nvme_pt_ioctl.c (hunk 3 above)
Comment (RevanthRajashekar) [reply src/lib/nvme_pt_ioctl.c]: Ohh my bad.. will fix these things before resubmitting with changes suggested.
Thanks!

File: src/lib/nvme_pt_ioctl.c (hunk 4 above)
Comment (RevanthRajashekar) [reply src/lib/nvme_pt_ioctl.c]: will do it

File: src/lib/nvme_pt_ioctl.c [hunk 6]
```diff
@@ -2006,53 +2015,73 @@ int opal_set_pwd_pt(struct sed_device *dev, enum SED_AUTHORITY auth, const struc
...
 	}
 
 	opal_dev = dev->priv;
 
-	if (mbr) {
-		enable_disable = OPAL_MBR_ENABLE;
-	} else {
-		enable_disable = OPAL_MBR_DISABLE;
-	}
+	if (mbr_done)
```
Comment (AndrzejJakowski) [inline src/lib/nvme_pt_ioctl.c]: you could make it one liner  which would look really pretty to me :)

File: src/lib/nvme_pt_ioctl.c [hunk 7]
```diff
@@ -2006,53 +2015,73 @@ int opal_set_pwd_pt(struct sed_device *dev, enum SED_AUTHORITY auth, const struc
...
-	ret = opal_set_mbr(dev->fd, opal_dev, OPAL_MBRDONE, en_dis);
+	ret = opal_set_mbr_done(dev->fd, opal_dev, enable_disable);
 	if (ret)
 		goto end_sessn;
 
 	ret = opal_end_session(dev->fd, opal_dev);
 	if (ret)
 		goto end_sessn;
 
-	ret = opal_start_admin1_lsp_session(dev->fd, opal_dev, &disk_key);
```
Comment (AndrzejJakowski) [inline src/lib/nvme_pt_ioctl.c:2051]: can this be handled in single session? 
why 2 sessions are started?

File: src/lib/nvme_pt_ioctl.c (hunk 5 above)
Comment (RevanthRajashekar) [reply src/lib/nvme_pt_ioctl.c]: will do it

File: src/lib/nvme_pt_ioctl.c (hunk 6 above)
Comment (RevanthRajashekar) [reply src/lib/nvme_pt_ioctl.c]: will do it

File: src/lib/nvme_pt_ioctl.c (hunk 7 above)
Comment (RevanthRajashekar) [reply src/lib/nvme_pt_ioctl.c:2051]: Thats correct, just a single session can handle this

**Comment by RevanthRajashekar** [comment]
Hey Andrzej,
Changes inculcated as suggested
Thanks!

File: src/lib/sed_ioctl.h [hunk 8]
```diff
@@ -29,9 +29,11 @@ int sedopal_setup_global_range(struct sed_device *dev, const struct sed_key *key
...
 			const char *user, enum SED_ACCESS_TYPE lock_type, uint8_t lr);
 
-int sedopal_shadowmbr(struct sed_device *dev, const char *password,
-		uint8_t key_len,
-		bool enable_mbr);
+int sedopal_shadowmbr(struct sed_device *dev, const struct sed_key *key,
+		      bool enable_mbr);
+
+int sedopal_mbrdone(struct sed_device *dev, const struct sed_key *key,
+                    bool mbr_done);
```
Comment (AndrzejJakowski) [inline src/lib/sed_ioctl.h]: ERROR: code indent should use tabs where possible

File: src/lib/nvme_pt_ioctl.c [hunk 9]
```diff
@@ -2006,53 +2015,59 @@ int opal_set_pwd_pt(struct sed_device *dev, enum SED_AUTHORITY auth, const struc
...
+	ret = opal_set_mbr_done(dev->fd, opal_dev, enable_disable);
 	if (ret)
 		goto end_sessn;
 
-	ret = opal_set_mbr_en_disable(dev->fd, opal_dev, en_dis);
+	ret = opal_set_mbr_en_disable(dev->fd, opal_dev, enable_disable);
```
Comment (AndrzejJakowski) [inline src/lib/nvme_pt_ioctl.c:2089]: why opal_set_mbr_done and opal_set_mbr_en_disable are both invoked here?
path through opal kernel driver just sets/unsets enable_diable

File: src/lib/nvme_pt_ioctl.c [hunk 10]
```diff
@@ -1528,7 +1526,14 @@ static int opal_write_datastr(int fd, struct opal_device *dev, const uint8_t *da
 	return ret;
 }
 
-static struct opal_req_item opal_read_datastr_cmd[] = {
+static int opal_write_datastr(int fd, struct opal_device *dev, const uint8_t *data,
```
Comment (AndrzejJakowski) [inline src/lib/nvme_pt_ioctl.c]: WARNING: line over 80 characters

File: src/lib/nvme_pt_ioctl.c [hunk 11]
```diff
@@ -1528,7 +1526,14 @@ static int opal_write_datastr(int fd, struct opal_device *dev, const uint8_t *da
...
+			      uint64_t offset, uint64_t size)
+{
+	return opal_generic_write_table(fd, dev, OPAL_DATASTORE_UID, data,
+					offset, size);
+}
+
+static struct opal_req_item opal_genric_read_table_cmd[] = {
```
Comment (AndrzejJakowski) [inline src/lib/nvme_pt_ioctl.c]: WARNING: 'genric' may be misspelled - perhaps 'generic'?

File: src/lib/nvme_pt_ioctl.c [hunk 12]
```diff
@@ -1561,28 +1568,28 @@ static int opal_read_datastr(int fd, struct opal_device *dev, uint8_t *data, uin
...
-		opal_read_datastr_cmd[3].val.uint = index + offset;
+		opal_genric_read_table_cmd[3].val.uint = index + offset;
 
 		len = MIN(OPAL_MAX_READ_TABLE, (end_row - index));
-		opal_read_datastr_cmd[7].val.uint = index + offset + len;
+		opal_genric_read_table_cmd[7].val.uint = index + offset + len;
 
-		prepare_req_buf(dev, opal_read_datastr_cmd, ARRAY_SIZE(opal_read_datastr_cmd),
-				opal_uid[OPAL_DATASTORE_UID], opal_method[OPAL_GET_METHOD_UID]);
+		prepare_req_buf(dev, opal_genric_read_table_cmd, ARRAY_SIZE(opal_genric_read_table_cmd),
```
Comment (AndrzejJakowski) [inline src/lib/nvme_pt_ioctl.c]: WARNING: line over 80 characters

File: src/lib/nvme_pt_ioctl.c [hunk 13]
```diff
@@ -1561,28 +1568,28 @@ static int opal_read_datastr(int fd, struct opal_device *dev, uint8_t *data, uin
...
+				opal_uid[table], opal_method[OPAL_GET_METHOD_UID]);
```
Comment (AndrzejJakowski) [inline src/lib/nvme_pt_ioctl.c]: WARNING: line over 80 characters

**Review by AndrzejJakowski — CHANGES_REQUESTED** [review]

File: src/lib/nvme_pt_ioctl.c (hunk 9 above)
Comment (RevanthRajashekar) [reply src/lib/nvme_pt_ioctl.c:2091]: will fix it

File: src/lib/sed_ioctl.h (hunk 8 above)
Comment (RevanthRajashekar) [reply src/lib/sed_ioctl.h]: fixed in commit #4

File: src/lib/nvme_pt_ioctl.c (hunk 9 above)
Comment (RevanthRajashekar) [reply src/lib/nvme_pt_ioctl.c:2089]: fixed in commit #4

File: src/lib/nvme_pt_ioctl.c (hunk 10 above)
Comment (RevanthRajashekar) [reply src/lib/nvme_pt_ioctl.c]: fixed in commit #4

File: src/lib/nvme_pt_ioctl.c (hunk 11 above)
Comment (RevanthRajashekar) [reply src/lib/nvme_pt_ioctl.c]: fixed in commit #4

File: src/lib/nvme_pt_ioctl.c (hunk 13 above)
Comment (RevanthRajashekar) [reply src/lib/nvme_pt_ioctl.c]: fixed in commit #4

**Comment by RevanthRajashekar** [comment]
Changes incorporated

**Review by AndrzejJakowski — APPROVED** [review]

**Pull request merged**